core/server/master/src/main/java/alluxio/master/block/BlockMasterWorkerServiceHandler.java

```diff
@@ -125,11 +131,24 @@ public void commitBlockInUfs(CommitBlockInUfsPRequest request,
   @Override
   public void getWorkerId(GetWorkerIdPRequest request,
       StreamObserver<GetWorkerIdPResponse> responseObserver) {
-    RpcUtils.call(LOG, (RpcUtils.RpcCallableThrowsIOException<GetWorkerIdPResponse>) () -> {
-      return GetWorkerIdPResponse.newBuilder()
-          .setWorkerId(mBlockMaster.getWorkerId(GrpcUtils.fromProto(request.getWorkerNetAddress())))
-          .build();
-    }, "getWorkerId", "request=%s", responseObserver, request);
+    if (request.hasClusterId()) {
+      // For compatibility, reuse the RPC of getWorkerID
+      // Determine whether it is preRegistration by the extendedRegisterInfo flag
+      String clusterId = request.getClusterId();
+      int blocksNum = request.getBlocksNum();
+      RpcUtils.call(LOG,
+          (RpcUtils.RpcCallableThrowsIOException<GetWorkerIdPResponse>) () -> {
+            // By the Flag, Let the Worker know that preRegister information is included
+            return mBlockMaster.workerPreRegister(
```

Comment:
is it possible to merge the logic into `mBlockMaster.getWorkerId()` call instead of having a separate method?

Reply:
Yes, but this may require changing the return type of `mBlockMaster.getWorkerId()`, Let it return `GetWorkerIdPResponse`

Reply:
@jiacheliu3  do you mean that:

alluxio.master.block.DefaultBlockMaster#getWorkerId
old
``` java
public long getWorkerId(WorkerNetAddress workerNetAddress)
```
new
```java
  public GetWorkerIdPResponse getWorkerId(String workerClusterId,
      WorkerNetAddress workerNetAddress, int blocksNum) throws IOException {
   //....
    PreRegisterCommandType commandType = checkWorker(curClusterId, workerClusterId, blocksNum);
    long workerId = getWorkerIdInternal(workerNetAddress); // the old getWorkerId rename to getWorkerIdInternal
    GetWorkerIdPResponse response = GetWorkerIdPResponse.newBuilder()
        .setPreRegisterCommandType(commandType)
        .setClusterId(curClusterId)
        .setWorkerId(workerId)
        .build();
    //...
    return response;
  }
}
```
the new getworkerid will contain logic for preRegister and old getWorkerId